BUG(request_sync_response) : add header validation for x-bsv-topic

cain4chain opened the pull request:
## Description of Changes

chore(x-bsv-topic) : add custom header for topic value and replace query params

## Linked Issues / Tickets

closes https://github.com/4chain-ag/go-overlay-services/issues/112

## Testing Procedure

chore(resquest_sync_response_test) : updated test cases using the header validation

- [ ] I have added new unit tests
- [ ] All tests pass locally
- [ ] I have tested manually in my local environment

## Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have run the linter

cain4chain committed:
BUG(request_sync_response) : add header validation for x-bsv-topic

chore(x-bsv-topic) : add custom header for topic value
chore(resquest_sync_response_test) : updated test cases using the header validation

github-actions[bot] committed:
chore: update generated files

mgosek-4chain commented on pkg/server/internal/app/commands/request_sync_response.go:
```diff
@@ -34,9 +34,9 @@ func (h *RequestSyncResponseHandler) Handle(w http.ResponseWriter, r *http.Reque
+	topic := r.Header.Get("x-bsv-topic")
 	if topic == "" {
-		http.Error(w, "missing 'topic' query parameter", http.StatusBadRequest)
+		http.Error(w, "missing 'x-bsv-topic' header", http.StatusBadRequest)
```
Please extract the string into an exported error type: 

```
// ErrMissingXBSVTopicHeader ....
var ErrMissingXBSVTopicHeader = errors.New("missing 'x-bsv-topic' header")
```

mgosek-4chain commented on pkg/server/internal/app/commands/request_sync_response_test.go:
```diff
@@ -49,7 +49,12 @@ func TestRequestSyncResponseHandler_Success(t *testing.T) {
 
 	// When:
-	resp, err := ts.Client().Post(ts.URL+"?topic=example-topic", "application/json", bytes.NewReader(body))
+	req, err := http.NewRequest(http.MethodPost, ts.URL, bytes.NewReader(body))
```
```
req, err := http.NewRequest(http.MethodPost, ts.URL, testutil.RequestBody(t, payload))
```

mgosek-4chain commented on pkg/server/internal/app/commands/request_sync_response_test.go:
```diff
@@ -73,7 +78,11 @@ func TestRequestSyncResponseHandler_MissingTopic(t *testing.T) {
 	// When:
-	resp, err := ts.Client().Post(ts.URL, "application/json", bytes.NewReader(body))
+	req, err := http.NewRequest(http.MethodPost, ts.URL, bytes.NewReader(body))
+	require.NoError(t, err)
```
```
req, err := http.NewRequest(http.MethodPost, ts.URL, testutil.RequestBody(t, payload))
```

mgosek-4chain commented on pkg/server/internal/app/commands/request_sync_response_test.go:
```diff
@@ -82,7 +91,7 @@ func TestRequestSyncResponseHandler_MissingTopic(t *testing.T) {
 	respBody, err := io.ReadAll(resp.Body)
 	require.NoError(t, err)
-	require.Contains(t, string(respBody), "missing 'topic'")
+	require.Contains(t, string(respBody), "missing 'x-bsv-topic'")
```
```
	actualErr := testutil.ParseToError(t, resp.Body)
	require.Equal(t, http.StatusBadRequest, resp.StatusCode)
	require.Equal(t, commands.ErrMissingXBSVTopicHeader, actualErr)
```

mgosek-4chain commented on pkg/server/internal/app/commands/request_sync_response_test.go:
```diff
@@ -93,7 +102,12 @@ func TestRequestSyncResponseHandler_InvalidJSON(t *testing.T) {
+	req, err := http.NewRequest(http.MethodPost, ts.URL, bytes.NewReader([]byte(`{invalid-json}`)))
+	require.NoError(t, err)
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("x-bsv-topic", "example-topic")
```
```
	req, err := http.NewRequest(http.MethodPost, ts.URL, testutil.RequestBody(t, `{invalid-json}`))
	require.NoError(t, err)

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-bsv-topic", "example-topic")
```

mgosek-4chain commented on pkg/server/internal/app/commands/request_sync_response_test.go:
```diff
@@ -117,7 +131,12 @@ func TestRequestSyncResponseHandler_InternalServerError(t *testing.T) {
+	req, err := http.NewRequest(http.MethodPost, ts.URL, bytes.NewReader(body))
+	require.NoError(t, err)
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set("x-bsv-topic", "example-topic")
```
```
	req, err := http.NewRequest(http.MethodPost, ts.URL, testutil.RequestBody(t, `{invalid-json}`))
	require.NoError(t, err)

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("x-bsv-topic", "example-topic")
```

I would recommend creating an exported const with defined header to avoid code duplication :)

cain4chain committed:
BUG (request_sync_response) : update as per review

chore(request_sync_response) : extract strings and export the error types
chore(request_sync_response_test) : add missing invalid method test, extract headers into function with optional bsv-topics

github-actions[bot] committed:
chore: update generated files
